src/lib/mina_net2/mina_net2.ml

```diff
@@ -1372,726 +121,419 @@ module Pubsub = struct
-                    | `Eof ->
-                        failwith "Eof"
-                    | `Ok q ->
-                        Base.Queue.peek_exn q
-                  in
-                  go (i - 1)
-              in
-              let%map () = go 1000 in
-              Pipe.write_without_pushback w "done" ;
-              Pipe.close w ;
-              handler_finished := true )
-          |> Deferred.Or_error.ok_exn
-        in
-        let%bind stream =
-          open_stream c ~protocol:"read_bytes" b_peerid >>| Or_error.ok_exn
-        in
-        let r, w = Stream.pipes stream in
-        for i = 0 to 999 do
-          Pipe.write_without_pushback w (Printf.sprintf "%d" i)
-        done ;
-        Pipe.close w ;
-        (* HACK: let our messages send before we reset.
-           It would be more principled to add flushing to
-           the stream interface. *)
-        let%bind () = after (Time.Span.of_sec 5.) in
-        let%bind _ = Stream.reset stream in
-        let%bind _msgs = Pipe.read_all r in
-        assert !handler_finished ;
-        shutdown ()
-      in
-      Async.Thread_safe.block_on_async_exn (fun () -> test_def)
-  *)
-
-    let%test_unit "stream" =
-      let () = Core.Backtrace.elide := false in
-      let test_def =
-        let open Deferred.Let_syntax in
-        let%bind b, c, shutdown = setup_two_nodes "test_stream" in
-        let%bind b_peerid = me b >>| Keypair.to_peer_id in
-        let handler_finished = ref false in
-        let%bind echo_handler =
-          handle_protocol b ~on_handler_error:`Raise ~protocol:"echo"
-            (fun stream ->
-              let r, w = Stream.pipes stream in
-              let%map () = Pipe.transfer r w ~f:Fn.id in
-              Pipe.close w ;
-              handler_finished := true)
-          |> Deferred.Or_error.ok_exn
-        in
-        let%bind stream =
-          open_stream c ~protocol:"echo" b_peerid >>| Or_error.ok_exn
-        in
-        let r, w = Stream.pipes stream in
-        Pipe.write_without_pushback w testmsg ;
-        Pipe.close w ;
-        (* HACK: let our messages send before we reset.
-           It would be more principled to add flushing to
-           the stream interface. *)
-        let%bind () = after (Time.Span.of_sec 1.) in
-        let%bind _ = Stream.reset stream in
-        let%bind msg = Pipe.read_all r in
-        (* give time for [a] to notice the reset finish. *)
-        let%bind () = after (Time.Span.of_sec 1.) in
-        let msg = Queue.to_list msg |> String.concat in
-        assert (String.equal msg testmsg) ;
-        assert !handler_finished ;
-        let%bind () = Protocol_handler.close echo_handler in
-        let%map () = shutdown () in
-        ()
+      match Hashtbl.find t.subscriptions subscription_id with
+      | Some (Subscription.E sub) ->
+          upon
+            (Subscription.handle_and_validate sub ~validation_expiration ~sender
+               ~data) (function
+            | `Validation_timeout ->
```

Comment:
Shall not we send `Ignore` validation at least?

Reply:
There is a timeout on the libp2p side that handles that already.